lib/web/sessions.go

```diff
@@ -497,148 +552,337 @@ func (s *sessionCache) ValidateTrustedCluster(validateRequest *auth.ValidateTrus
 	return s.proxyClient.ValidateTrustedCluster(validateRequest)
 }
 
-func (s *sessionCache) InvalidateSession(ctx *SessionContext) error {
-	defer ctx.Close()
-	if err := s.resetContext(ctx.GetUser(), ctx.GetWebSession().GetName()); err != nil {
-		return trace.Wrap(err)
+// validateSession validates the session given with user and session ID.
+// Returns a new or existing session context.
+func (s *sessionCache) validateSession(ctx context.Context, user, sessionID string) (*SessionContext, error) {
+	sessionCtx, err := s.getContext(user, sessionID)
+	if err == nil {
+		return sessionCtx, nil
 	}
+	if !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+	return s.newSessionContext(user, sessionID)
+}
+
+func (s *sessionCache) invalidateSession(ctx *SessionContext) error {
+	defer ctx.Close()
 	clt, err := ctx.GetClient()
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	err = clt.DeleteWebSession(ctx.GetUser(), ctx.GetWebSession().GetName())
-	return trace.Wrap(err)
+	// Delete just the session - leave the bearer token to linger to avoid
+	// failing a client query still using the old token.
+	err = clt.WebSessions().Delete(context.TODO(), types.DeleteWebSessionRequest{
+		User:      ctx.user,
+		SessionID: ctx.session.GetName(),
+	})
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+	if err := s.resetContext(ctx.GetUser(), ctx.session.GetName()); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
 }
 
-func (s *sessionCache) getContext(user, sid string) (*SessionContext, error) {
-	s.Lock()
-	defer s.Unlock()
-
-	val, ok := s.contexts.Get(user + sid)
+func (s *sessionCache) getContext(user, sessionID string) (*SessionContext, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	ctx, ok := s.sessions[user+sessionID]
 	if ok {
-		return val.(*SessionContext), nil
+		return ctx, nil
 	}
-	return nil, trace.NotFound("sessionContext not found")
+	return nil, trace.NotFound("no context for user %v and session %v",
+		user, sessionID)
 }
 
-func (s *sessionCache) insertContext(user, sid string, ctx *SessionContext, ttl time.Duration) (*SessionContext, error) {
-	s.Lock()
-	defer s.Unlock()
+func (s *sessionCache) insertContext(user string, ctx *SessionContext) (exists bool) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	id := user + ctx.session.GetName()
```

Comment:
I'd consider making a function, `makeKey` or something, that implements this concatenation logic since it's used in a few places.